libkbfs/quota_usage.go

```diff
@@ -70,18 +77,69 @@ func NewEventuallyConsistentTeamQuotaUsage(
 	return q
 }
 
+func (q *EventuallyConsistentQuotaUsage) getCached() cachedQuotaUsage {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+	return q.cached
+}
+
+func (q *EventuallyConsistentQuotaUsage) getID(
+	ctx context.Context) (keybase1.UserOrTeamID, error) {
+	if q.tid.IsNil() {
+		session, err := q.config.KBPKI().GetCurrentSession(ctx)
+		if err != nil {
+			return keybase1.UserOrTeamID(""), err
+		}
+		return session.UID.AsUserOrTeam(), nil
+	}
+	return q.tid.AsUserOrTeam(), nil
+}
+
 func (q *EventuallyConsistentQuotaUsage) getAndCache(
 	ctx context.Context) (err error) {
 	defer func() {
 		q.log.CDebugf(ctx, "getAndCache: error=%v", err)
 	}()
+
+	// Try pulling the quota from the disk cache.  If it exists, still
+	// try the servers, but give it a short timeout.
+	var quotaInfoFromCache *kbfsblock.QuotaInfo
+	id, err := q.getID(ctx)
+	if err != nil {
+		return err
+	}
+	getCtx := ctx
+	dqc := q.config.DiskQuotaCache()
+	if dqc != nil {
+		qi, err := dqc.Get(ctx, id)
+		if err == nil {
+			q.log.CDebugf(ctx, "Read quota for %s from disk cache", id)
+			quotaInfoFromCache = &qi
+			var cancel context.CancelFunc
+			getCtx, cancel = context.WithTimeout(
+				ctx, quotaServerTimeoutWhenCached)
+			defer cancel()
+		}
+	}
+
 	var quotaInfo *kbfsblock.QuotaInfo
 	if q.tid.IsNil() {
-		quotaInfo, err = q.config.BlockServer().GetUserQuotaInfo(ctx)
+		quotaInfo, err = q.config.BlockServer().GetUserQuotaInfo(getCtx)
 	} else {
-		quotaInfo, err = q.config.BlockServer().GetTeamQuotaInfo(ctx, q.tid)
+		quotaInfo, err = q.config.BlockServer().GetTeamQuotaInfo(getCtx, q.tid)
 	}
-	if err != nil {
+	doCacheToDisk := dqc != nil
+	switch err {
+	case nil:
+	case context.DeadlineExceeded:
+		if quotaInfoFromCache != nil {
+			q.log.CDebugf(ctx, "Can't contact server; using cached quota")
+			quotaInfo = quotaInfoFromCache
+			doCacheToDisk = false
+		} else {
+			return err
+		}
+	default:
```

Comment:
In offline mode, we'll probably hit this `default` case right? Seems we shouldn't return error in that case.

Reply:
I don't think so, we should hit the `DeadlineExceeded` case in offline mode.  Remember the RPC layer just waits for the connection to come up, it doesn't return any other TCP disconnected errors.